src/libsyntax/ext/expand.rs

```diff
@@ -1366,51 +1364,25 @@ impl<'a, 'b> Folder for InvocationCollector<'a, 'b> {
                 self.cx.current_expansion.directory_ownership = orig_directory_ownership;
                 result
             }
-            // Ensure that test functions are accessible from the test harness.
+
+            // Ensure that test items can be exported by the harness generator.
             // #[test] fn foo() {}
             // becomes:
             // #[test] pub fn foo_gensym(){}
-            // #[allow(unused)]
-            // use foo_gensym as foo;
-            ast::ItemKind::Fn(..) if self.cx.ecfg.should_test => {
-                if self.tests_nameable && item.attrs.iter().any(|attr| is_test_or_bench(attr)) {
-                    let orig_ident = item.ident;
-                    let orig_vis   = item.vis.clone();
-
+              ast::ItemKind::Const(..)
+            | ast::ItemKind::Static(..)
+            | ast::ItemKind::Fn(..) if self.cx.ecfg.should_test => {
+                if self.tests_nameable && attr::contains_name(&item.attrs, "test_case") {
                     // Publicize the item under gensymed name to avoid pollution
+                    // This means #[test_case] items can't be referenced by user code
                     item = item.map(|mut item| {
                         item.vis = respan(item.vis.span, ast::VisibilityKind::Public);
```

Comment:
This transformation should ideally be moved into the `#[test]` macro as well.

Reply:
If we want to pursue https://github.com/rust-lang/rust/pull/53410#issuecomment-417865628.

Reply:
Hmm moving the transform to `#[test]` would mean that we would need to have a lint for all `#[test_case]` items to be pub, because they need to be for reexporting to work.

I could make `#[test_case]` another syntax extension that performs that transform and use a third inert, internal-only attribute to mark tests. It seems a bit much but it would clear up the "random stuff in `expand.rs`" problem. Thoughts?

Reply:
Hmmm is there a good way to gate built-in extensions? I'm unsure how I would issue the proper error message if you tried to use `#[test_case]` without `#![feature(custom_test_frameworks)]`

Reply:
Yes, see for example, how `global_asm!` is gated - https://github.com/rust-lang/rust/blob/master/src/libsyntax_ext/global_asm.rs

Reply:
Gotcha. One more (hopefully last) concern. If we make `#[test_case]` into a syntax extension, then it could be shadowed in the expansion context of `#[test]`. Is there a way to directly address builtin attributes?

Reply:
Or should I just manually call `test_case`'s expand function from within `test`'s?

Reply:
>Is there a way to directly address builtin attributes?

Not right now (built-in types have the [same problem](https://github.com/rust-lang/rust/issues/44865)).
Longer term solution is to switch built-in syntax extensions to hygienic-by-default, it will cut off all unnecessary interactions like this.

Reply:
>Or should I just manually call test_case's expand function from within test's?

That would work.